server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java

```diff
@@ -212,6 +214,78 @@ public void onFailure(Exception e) {
+        repositoriesService.applyClusterState(new ClusterChangedEvent("put unstable repository", clusterState, emptyState()));
+
+        var repo = repositoriesService.repository(repoName);
+        assertThat(repo, isA(InvalidRepository.class));
+
+        clusterState = createClusterStateWithRepo(repoName, TestRepository.TYPE);
+        repositoriesService.applyClusterState(new ClusterChangedEvent("put test repository", clusterState, emptyState()));
+        repo = repositoriesService.repository(repoName);
+        assertThat(repo, isA(TestRepository.class));
+    }
+
+    // test remove InvalidRepository when current repo is removed in cluster state
+    public void testRemoveInvalidRepositoryTypeWhenApplyingClusterState() {
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+
+        var clusterState = createClusterStateWithRepo(repoName, UnstableRepository.TYPE);
+        repositoriesService.applyClusterState(new ClusterChangedEvent("put unstable repository", clusterState, emptyState()));
+        repositoriesService.applyClusterState(new ClusterChangedEvent("removing repo", emptyState(), clusterState));
+
+        assertThatThrows(
+            () -> repositoriesService.repository(repoName),
+            RepositoryMissingException.class,
+            equalTo("[" + repoName + "] missing")
+        );
+    }
+
+    // InvalidRepository is created when current node is non-master node and failed to create repository by applying cluster state from
+    // master. When current node become master node later and same repository is put again, current node can create repository successfully
+    // and replace previous InvalidRepository
+    public void testRegisterRepositorySuccessAfterCreationFailed() {
+        // 1. repository creation failed when current node is non-master node and apply cluster state from master
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+
+        var clusterState = createClusterStateWithRepo(repoName, UnstableRepository.TYPE);
+        repositoriesService.applyClusterState(new ClusterChangedEvent("put unstable repository", clusterState, emptyState()));
+
+        var repo = repositoriesService.repository(repoName);
+        assertThat(repo, isA(InvalidRepository.class));
+
+        // 2. repository creation successfully when current node become master node and repository is put again
+        var request = new PutRepositoryRequest().name(repoName).type(TestRepository.TYPE);
+
+        repositoriesService.registerRepository(request, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                assertTrue(acknowledgedResponse.isAcknowledged());
+                assertThat(repositoriesService.repository(repoName), isA(TestRepository.class));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("[" + repoName + "] repository type [test] faield to create");
```

Comment:
Suggest failing like this so we see the cause:

```suggestion
                assert false : e;
```